### ISSUE TITLE
getAccessTokenSilently redirects to the callback URL on refresh token expiration (login_required error)

### ISSUE THREAD

**Issue by nstanard** [auth0/auth0-react#354]
### Describe the problem you'd like to have solved

> getAccessTokenSilently redirects to the callback URL after throwing "login_required" errors. (after refresh token expiration)

There are a few things that are frustrating about that. 

1. The getTokenSilently method in the auth0-spa-js package appears to not redirect. This caused me a good bit of confusion at first because it wasn't clear to me that there was a difference between that call and this one.
2. I can't use a try/catch to handle this in any special or custom way. (I specifically need to invalidate the application cookie and do some cleanup actions _before_ I send the user to the login screen.)
3. The code from `auth0-spa-js` even says "There's no actual redirect when getting a token silently," when looking at the option object. That statement isn't true for the getAccessTokenSilently from this package. (this auth0-react package uses the auth0-spa-js package though)

<img width="1746" alt="image" src="https://user-images.githubusercontent.com/4204806/164464940-545b6354-fff3-4a3a-970f-dc809ab0b22e.png">

**Please make it so I can pass in an option to get token silently so it doesn't redirect automatically.**  🙏 

https://community.auth0.com/t/getaccesstokensilently-redirects-to-login-when-session-has-expired/74541
This is a related community post ^

https://community.auth0.com/t/getaccesstokensilently-throws-error-login-required/52333
This post is inaccurate and misleading ^

**Comment by adamjmcgrath** [comment]
Hi @nstanard - thanks for raising this

> getAccessTokenSilently redirects to the callback URL after throwing "login_required" errors.

Where did you see this statement? This SDK won't redirect you anywhere after after getting a `login_required` error. It will just clear your authenticated state (the `user` and `isAuthenticated` properties).

**Comment by nstanard** [comment]
> Hi @nstanard - thanks for raising this
> 
> > getAccessTokenSilently redirects to the callback URL after throwing "login_required" errors.
> 
> Where did you see this statement? This SDK won't redirect you anywhere after after getting a `login_required` error. It will just clear your authenticated state (the `user` and `isAuthenticated` properties).

This isn't a statement I read anywhere. This is what's happening in the application I'm integrating at the moment. I have the call to getAccessTokenSilently wrapped in a try/catch. I have refresh tokens turned on and I'm actively testing the access token expiration (token rotation) as well as the refresh token expiration (login_required error). When the refresh token is active the call to `getAccessTokenSilently` works as expected and I get the token back. However, when I unlink my refresh token or let it expire (after the 5 min window I set in the configs) the call to `getAccessTokenSilently` throws an error saying "Login Required". Inside the catch block I am trying to handle the error manually but before I can do anything the auth0-react package redirects the user/location to the callback URL. 

Unfortunately I don't think I can share the repo since it's private.

**Comment by nstanard** [comment]
@adamjmcgrath This seems related: https://github.com/auth0/auth0-spa-js/blob/0ae6078c02d3a4184147e788adfceae4db78db65/src/Auth0Client.ts#L1220

What are the chances the call to `oauthToken` is using the redirect info being passed in?

**Comment by nstanard** [comment]
Yes! You are absolutely correct. I am using the HOC `withAuthenticationRequired` and I realize now that is what's forcing the user to the callback url or the login UI based on the isAuthenticated status. 

I must say though that the examples about catching the `login_required` errors really are misleading when a person is using `withAuthenticationRequired`. 

Please correct me if I'm wrong but refreshing the page with `withAuthenticationRequired` will always cause the iframe redirect logic, yes? There's no way to get the auth0-react package to persist `isAuthenticated` without using `withAuthenticationRequired` ?

**Comment by adamjmcgrath** [comment]
> Please correct me if I'm wrong but refreshing the page with `withAuthenticationRequired` will always cause the iframe redirect logic, yes?

I'm not sure what you mean by "iframe redirect logic", but it will always redirect the browser to the login page if the SDK has initialised and the user is not logged in

> There's no way to get the auth0-react package to persist `isAuthenticated` without using `withAuthenticationRequired`?

`isAuthenticated` will behave the same regardless of if you use `withAuthenticationRequired`, there's no requirement to use it. You should only use it if you want the behaviour I described in the previous answer (to always redirect the browser to the login page if the SDK has initialised and the user is not logged in)

**Comment by adamjmcgrath** [comment]
Closing - I think this has answered your question @nstanard - ping me if you want to reopen it

**Comment by nstanard** [comment]
This has answered my question @adamjmcgrath and thank you for your time. I must say though that `isAuthenticated` was not persisting for me in every situation before using `withAuthenticatedRequired`. I think there's room for improvement for how the Auth0-React package is documented in relation to the Auth0 tenant/application/api configuration.

**Comment by adamjmcgrath** [comment]
Thanks for the feedback @nstanard 

> I must say though that isAuthenticated was not persisting for me in every situation before using withAuthenticatedRequired

If you can create a reproducible example of this I'd be happy to investigate

**Comment by nstanard** [comment]
@adamjmcgrath  

Our app authenticates with Auth0 then uses the silent token to authenticate with our API. So our app sessions are really managed by our internal API and not directly by Auth0. That means that I have the option to not use `withAuthenticationRequired` which I've opted for. I am still trying to figure out what all determines if `isAuthenticated` gets persisted between page refreshes. I'll continue to test and investigate. One thing I noticed that has an impact on the `isAuthenticated` flag persisting as `true` is the `cachelocation` setting being `memory` vs `localstorage`.
